### PR TITLE
Grid Inventory

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

cooltext448060731009762

Based on [this accepted proposal.](https://github.com/space-wizards/docs/pull/99)

Changes storage to use a grid-based. 
Storage containers, instead of lists, are now a single fixed grid above the player's hands. Items are spatially represented as rectangles (with support for other non-rectangular shapes) and can be freely rotated and moved throughout the grid.

All current storage and item balance is being (mostly) preserved, as the area of the grids and item shapes corresponds to the max values of storage containers and items' weight, respectively. The main difference is simply the presentation and the ability for the configuration of items within a container to block the insertion of other items.

The system additionally has an auto-insert feature, used for storage fills and basic insertion, which automatically places the item in a valid location on the grid.

## Media

image
_example backpack_

image
_the default survival box_

image
_the china lake bundle_

image
_a demonstration of the different item sizes and their default shapes_

[Content.Client_DaXiHbPWkl.webm](https://github.com/space-wizards/space-station-14/assets/98561806/b8f151b1-1ad6-485b-b118-63fc141b7ef0)

_a demonstration of the auto-insertion, dragging, rotation, and blocking of items_

https://github.com/space-wizards/space-station-14/assets/98561806/eb35f530-9db6-493e-93b9-ead9b6b67a05

_demo of direct insertion from the hand_

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

Custom ItemSizePrototypes now need to define a default shape.
All instances of StorageComponent need to replace `MaxTotalWeight` with the appropriate value for `StorageGrid`. The easiest way to do this is to create a rectangle with an area equal to your `MaxTotalWeight`, which should mostly retain similar balance.

All of the UI code for storage is obviously completely different.

**Changelog**

:cl:
- add: Storage now uses a grid-based system. You can drag around items in your bag, left click to remove them, and use right click to rotate them while dragging.
- add: The storage window is now a fixed singleton above the hands. You can make it moveable by toggling "Static storage UI" in the controls menu.

### CODE REVIEW

**Comment by DanSAussieITS** [comment]
The madlad actually did it.

It would be nice to see certain items having fixed orientations relative to their grid shape if possible. Main one that sticks out in the video is the screwdriver when oriented vertically.

**Comment by Boaz1111** [comment]
> The madlad actually did it.
> 
> 
> 
> It would be nice to see certain items having fixed orientations relative to their grid shape if possible. Main one that sticks out in the video is the screwdriver when oriented vertically.

I agree, the sprite sizes and orientations could do with some work... might be hard to pull off though.

**Comment by GreyMario** [comment]
part of the reason ui windows are movable is so that you can put them in places where they won't obstruct things you need to be able to see. i don't agree with having this be a fixed position ui glued to your item bar, especially since it's beginning to encroach upon the middle of the screen _rather severely_

i also don't agree with allowing only one inventory to be open; the resulting ui clutter does counterbalance any imagined powergaming "advantage"

**Comment by DiaAWAY** [comment]
Not a huge fan of having to use right mouse button to pick stuff up, i think that dragging should be "main button" and then holding it in for a certain amount of time. I believe this is very ordinary behaviour on most grid inventories.

**Comment by DiaAWAY** [comment]
This might be a minor bug, but when i open the backpack, and then try to open the survival box (both times i use the e button), the survival box inventory opens then instantly closes.

**Comment by EmoGarbage404** [comment]
> This might be a minor bug, but when i open the backpack, and then try to open the survival box (both times i use the e button), the survival box inventory opens then instantly closes.

Im aware and the PR isn't even close to finished yet.

**Comment by EmoGarbage404** [comment]
> Not a huge fan of having to use right mouse button to get stuff from inventory, I think that dragging should be whatever main button people choose, and then holding it in for a certain amount of time. I believe this is very ordinary behaviour on most grid inventories.

I think you misunderstood. Dragging is in fact the main button (left click) and right click removes the item from the storage into your hand.

**Comment by EmoGarbage404** [comment]
> Right now it feels more jarring to have the inventory open and having an item in hand, and not get a hover overlay in the open inventory window and be able to place the item into the inventory with my mouse.

Yeah, i think that might feel better. I'll mess around with it.

**Comment by asperger-sind** [comment]
When you can't insert an item it should probably tell you the size proportions of the item e.g "No room! (UristMcItem is 4x4)"

also generic "can you please allow moving windows" request

**Comment by DiaAWAY** [comment]
> > Right now it feels more jarring to have the inventory open and having an item in hand, and not get a hover overlay in the open inventory window and be able to place the item into the inventory with my mouse.
> 
> Yeah, i think that might feel better. I'll mess around with it.

For sure! 

Also when using left mouse button and trying to rotate an item it it feels awkward to have to push both buttons at the same time (at least it does for me, i use middle mouse button to rotate things), 

Maybe have it so that when you pick up the item, it stays picked up so you can rotate and place it in peace?
The previous, old slot, could then have a ghost version of the item visible for indication that it isn't moved yet - that way can you can close the inventory, press escape or other actions that cancel\abort the action so you're not stuck with a coca cola on your cursor\hand if a nukie comes blasting around the corner :)

My 5 cents for now!

**Comment by EmoGarbage404** [comment]
Rotation is bound to middle mouse by default, so im not exactly sure what your issue is.

As for the dragging, the item actually always remains in the slot until you place it down (at which point it is moved). Its not visible while you're dragging it, but it's still there.

**Comment by EmoGarbage404** [comment]
> i don't agree with having this be a fixed position ui glued to your item bar, especially since it's beginning to encroach upon the middle of the screen _rather severely_

I mean, at the reasonable UI scale I have it at, I would hardly call it severe. It's barely reaching a quarter of the screen's vertical space when fully extended. And even still, it's not an omnipresent HUD element. You can easily close and reopen it whenever. 

If you don't like how much space having a window open takes up on your screen, You can just close it and only open it when you need to retrieve an item. The quick insert/retrieval keys have existed for a while for pretty much this sole purpose. Nothing in the game dictates that you _must_ be able to grab any item from your person instantly.

> the resulting ui clutter does counterbalance any imagined powergaming "advantage"

I mean this just really isn't true? It is extremely common to see screenshots of people with 3+ windows open to allow them to instantly grab any item they want from any bag. To imply that having a few small boxes in the black space on the edge of the screen is an actual counterbalance to being able to _instantly grab any item from any container on your person_ is frankly laughable.

**Comment by deltanedas** [comment]
> Nothing in the game dictates that you must be able to grab any item from your person instantly.

you can still do this its just you cant see 5% of the screen

**Comment by asperger-sind** [comment]
> > Nothing in the g...
> you can still do this its just you cant see 5% of the screen

It's more than 5% on low-res monitors, which is part of the reason why I want to see movable windows.

**Comment by GreyMario** [comment]
> Nothing in the game dictates that you _must_ be able to grab any item from your person instantly.

>  To imply that having a few small boxes in the black space on the edge of the screen is an actual counterbalance to being able to _instantly grab any item from any container on your person_ is frankly laughable.

please explain why inventory management MUST be a chore, because that seems to be the argument you're insistent on making

**Comment by metalgearsloth** [comment]
> > > Nothing in the g...
> > > you can still do this its just you cant see 5% of the screen
> 
> It's more than 5% on low-res monitors, which is part of the reason why I want to see movable windows.

UI has a scaling option for it.

> please explain why inventory management MUST be a chore

It's not a chore and being able to keep all inventory windows on screen at minimum size is unintended behaviour that was getting fixed one way or another.

**Comment by EmoGarbage404** [comment]
> please explain why inventory management MUST be a chore, because that seems to be the argument you're insistent on making

please acknowledge that there's some middle ground between "completely unrestricted" and "a chore."

**Comment by deathride58** [comment]
As we've said on the Discord, the ease of opening and closing a UI does *not* at all justify actively regressing UX by removing the simple ability to move the storage UI. It's genuinely disappointing to us that the UI's been kept static despite the conversation, as we've directly pointed out that making it static makes it *far* clunkier for both the average user and the typical powergamer (notably: making the storage UI static directly benefits those who have their rotate viewport binds set to accessible keys, which the average user probably does *not*. We're pretty confident that everyone can see why it'd be pretty awful to see the rotate viewport binds end up potentially being considered a necessity among players). We feel like our previously stated concerns have been outright ignored.

Additionally, there's particularly major accessibility concerns with having the UI be static. Small screens (or anyone who otherwise relies on larger UI scaling values, most importantly being players with low vision) in particular are the *biggest* point you need to consider, especially with SS14 being performant enough to run on devices that're traditionally locked to smaller screens, such as laptops. Keeping the storage UI movable directly accommodates these players, as the ability to shunt the storage UI to the side of the screen would do wonders to alleviate the massive discrepancy in how much visibility can be affected across a variety of screen sizes and UI scales.

It doesn't seem like an unreasonable request at all for the storage UI to remain movable, regardless of whether or not the storage UI becomes a singleton. The benefits of keeping it as such genuinely outweighs all of the positives we can potentially see with making it static (especially since making it static is both a UX regression and a potential accessibility problem).

**Comment by BlazeBa** [comment]
What about bluespace bags? How will they even work? Should I close the entire player screen? Or will there be a scroll for very large bags?

**Comment by DiaAWAY** [comment]
> Rotation is bound to middle mouse by default, so im not exactly sure what your issue is.
> 
> As for the dragging, the item actually always remains in the slot until you place it down (at which point it is moved). Its not visible while you're dragging it, but it's still there.

Hmm, you don't feel its awkward to have to hold down the drag mouse button, then while holding it down (so it doesn't release the item) click the middle mouse button? 

The default rotate button in SS14 is, as far as i know, used "stand alone" without having to hold any other buttons down simultaneously. 
The build menu in SS14 doesn't feel awkward to me since the rotation button is only done after you have clicked something in the build menu itself.

**Comment by deathride58** [comment]
To give extra weight as to why it's a bad idea to have item rotation be hard-coded to be bound to MMB: The middle mouse button is *not* universally ergonomic at all (not to mention that there's quite a few mice that simply don't have an MMB at all). To illustrate, here's the mouse we have:
image

(For posterity, isn't just a trackball thing; there's quite a few ergonomic mice out there with the scroll wheel and left mouse button both intended to be used exclusively by the thumb)
In order to press MMB while holding down LMB, we have to either curl our index finger down in a genuinely painful manner, or use our other hand just to press MMB. Both of these options are frankly completely unreasonable from a UX standpoint. (And neither is "just use a normal mouse"; finger trackballs are the only mice that we've found to be reasonably compatible with our arthritis.)

Being able to rebind the button to rotate an item would make this a *complete* nonissue. This button should not at all be hardcoded.

**Comment by EmoGarbage404** [comment]
yeah i can add in custom binds for all the shit that's not a huge issue.

**Comment by deltanedas** [comment]
now imagine trackpad enjoyers having a stroke just trying to do basic inventory management :trollface: mmb and lmb at same time is not possible on some of them

**Comment by EmoGarbage404** [comment]
> now imagine trackpad enjoyers having a stroke just trying to do basic inventory management :trollface: mmb and lmb at same time is not possible on some of them

and this is why rebinding it is being implemented.

**Comment by OtherSkittle** [comment]
Screwing over the average player is ok because we get to mildly inconvenience powergamers for a day before they set up new macros and then they are no longer affected at all.

I don't even understand what advantage this purposefully horrible UX is supposed to even prevent in the first place? If anything it gives an advantage to players that are willing and able to mess around with their screen positions to make it so it doesn't cover any visible playing space at all, and if you aren't in this group of people (and I think most people aren't) you just get worse UX for no reason at all.

**Comment by OtherSkittle** [comment]
> If you don't want to have a polite discussion then don't comment.

There is no good reason for why the inventory window cannot be moved and the author has already basically ignored every attempt from others to claim this fact so I'm not exactly sure how I am supposed to be creating a discussion on this topic with the author.

Obviously the PR is a work in progress so it cannot really be judged at the moment but considering the author's only response to the topic of static windows has been to imply that it's a feature not a bug I don't really know what to say.

Personally I feel that static windows fail to "fix" 'people having instant access to every item in their inventory' because it doesn't stop players with macros from continuing to do just that while the average player gets screwed over by 2 extra interactions they have to do for every bit of non hotkey inventory management they are doing just so they can see the bottom part of their screen (or as a previous person mentioned start abusing the rotation keys both of these being things that I doubt the average player would be happy to do).

And on a final note I will repeat my question from my previous message: The fact that the UX is bad has been implied by the author to be a feature so I just want to know what exactly are powergamers doing that refusing to add movable windows is fixing? (since as I said before it isn't fixing instant access to inventory imo)

**Comment by IProduceWidgets** [comment]
This is looking awesome. I know you mentioned non-rectangular shapes will be possible, but I wanted to throw my 2c in for making sure it can be *really weird* shapes and not just your standard tetris pieces.

**Comment by EmoGarbage404** [comment]
> This is looking awesome. I know you mentioned non-rectangular shapes will be possible, but I wanted to throw my 2c in for making sure it can be _really weird_ shapes and not just your standard tetris pieces.

image

this good enough?

**Comment by deathride58** [comment]
**Please** at the very least acknowledge the UX and accessibility concerns that we and several other folks have brought up here. It's genuinely distressing to see an otherwise excellent PR end up with some criticism (a major part of our own dating back to about a week ago) ending up outright ignored, whilst other drops of criticism end up not only addressed, but resolved outright in under half an hour, without even the slightest hint of why the other criticism isn't even being acknowledged.

The end-user experience is *crucial* for major systems like this to sit well with a game's playerbase. We genuinely don't know what we have to say for our expertise to even be heard here. Please don't let this PR end up a disaster through negligence. We are actively begging you.

**Comment by Brandon-Huu** [comment]
How will this work with [ SHIFT ] + [ B ] ?

**Comment by SonicHDC** [comment]
> How will this work with [ SHIFT ] + [ B ] ?

Moves item to first empty space from left top corner

**Comment by Anzarot121** [comment]
And so, all the previous inventory changes aimed at fighting power gamers and munchkins had no effect on them, so you decided to roll back the inventory to at least the era of the first Stalker. Are you sure they won't find a way to turn this into an advantage against regular players? For example, with the current system, I regularly see bags full of combat medkits because they functionally double the inventory size. However, most likely you will ignore my question, just as you ignored all previous questions on this topic. 
Sloth, I’m not being offensive, Google Translator simply does not convey the necessary intonations or conveys unnecessary ones.

**Comment by metalgearsloth** [comment]
> Sloth, I’m not being offensive,

> And so, all the previous inventory changes aimed at fighting power gamers and munchkins had no effect on them, so you decided to roll back the inventory

> However, most likely you will ignore my question

I don't know how these would be any nicer in whatever other language.

> Please at the very least acknowledge the UX and accessibility concerns that we and several other folks have brought up here. 

As far as I'm aware the only request made was moveable / resizeable window given this will be somewhat larger than the slot-inventory from ss13.

@EmoGarbage404 I think another case to be made is the fact that with the inventory bar + storage open it encroaches a lot of southern space on your screen which might unnecessarily favour attacking from bottom -> top rather than vice versa in which case being able to move it would negate any sort of advantage.

Part of the reason we use a square viewport is so people don't just rotate their camera to get the optimal viewport with widescreen.

**Comment by EmoGarbage404** [comment]
I don’t know why this is being framed like something intended solely for killing powergamers.

First and foremost, the changes are to make storage just a more interesting and dynamic process than the frankly placeholder system that it is currently. Extending factors like limited windows is simply for external balancing factors.

Considering that bhijin has made a good point for it, I'll look into making the main window moveable (with the conceit that it might take a while and be a pain in my ass :P)

**Comment by deathride58** [comment]
From the bottom of our heart: thank you.

**Comment by BolloTea** [comment]
I don't see any problems with statics.  if people are not comfortable playing with equipment like in stalker or eft, then this is solely a problem and a problem of their adaptability.  I'm not trying to insult the players, but I just want them to direct these emotions to improve their skills, and not look for easy ways

**Comment by Anzarot121** [comment]
> I don’t know why this is being framed like something intended solely for killing powergamers.
Because this was one of Sloth’s main arguments for the previous redesign of the storage system. The other was the SS13 "tradition", which in SS13 is due to engine limitations.
https://github.com/space-wizards/space-station-14/issues/21134#issuecomment-1773592258
This will not help the players in any way, it will increase the entry threshold, this is either done to combat the manchins, or it is simply complicating things for the sake of complicating them, which is even worse. I don't see any other options.

**Comment by OtherSkittle** [comment]
> I don't see any problems with statics. if people are not comfortable playing with equipment like in stalker or eft, then this is solely a problem and a problem of their adaptability. I'm not trying to insult the players, but I just want them to direct these emotions to improve their skills, and not look for easy ways

Tarkov is not SS14 and they do not seek to provide similar experiences. Not only that but this system does not emulate Tarkov's inventory beyond the most bare similarities (grid based and that's pretty much it). And to end this aside for pretty much forever I hope, Tarkov has movable inventory windows, which is the entire point of this discussion.


> I don’t know why this is being framed like something intended solely for killing powergamers.

Your single response on the matter was framed on the basis that windows shouldn't be movable because

> Nothing in the game dictates that you must be able to grab any item from your person instantly.

which either means that the target is powergamers that "abuse" the current system by "having instant access to any item on their person", OR you're attempting to make the inventory management more realistic which the current PR + future planned changes fail to emulate so doubt that is the reason. If you were willing to respond to people on this subject then maybe people wouldn't keep getting the wrong idea.

> First and foremost, the changes are to make storage just a more interesting and dynamic process than the frankly placeholder system that it is currently. Extending factors like limited windows is simply for external balancing factors.

Nobody is saying that the PR's only purpose is to attack powergamers. The entire discussion was centered around the movability of the inventory windows, which I hardly think counts as an external balancing factor unless you're willing to agree that the purpose of your refusal to change it (in previous comments) was to balance the game against people who abuse it to "have instant access to every item".

If the purpose of the PR is to replace the current system then you shouldn't be surprised when people complain that parts of it are strictly worse for no stated reason.

**Comment by Zumorica** [comment]
This is really cool, great job!

**Comment by DiaAWAY** [comment]
> > Sloth, I’m not being offensive,
> 
> > And so, all the previous inventory changes aimed at fighting power gamers and munchkins had no effect on them, so you decided to roll back the inventory
> 
> > However, most likely you will ignore my question
> 
> I don't know how these would be any nicer in whatever other language.
> 
> > Please at the very least acknowledge the UX and accessibility concerns that we and several other folks have brought up here.
> 
> As far as I'm aware the only request made was moveable / resizeable window given this will be somewhat larger than the slot-inventory from ss13.
> 
> @EmoGarbage404 I think another case to be made is the fact that with the inventory bar + storage open it encroaches a lot of southern space on your screen which might unnecessarily favour attacking from bottom -> top rather than vice versa in which case being able to move it would negate any sort of advantage.
> 
> Part of the reason we use a square viewport is so people don't just rotate their camera to get the optimal viewport with widescreen.

this is off-topic; but the square viewport thing, is that currently implemented? because i can definitively rotate my camera to see further. Even Liltenhead does it in his latest video.

**Comment by EmoGarbage404** [comment]
Fyi you can make it square just by changing a cvar. It's something I'm aware of and intended to change.

**Comment by MACMAN2003** [comment]
holy shit deus ex inventory....... ss14 is one step closer to becoming a true immersive sim.
will there be an auto-sort button just in case someone doesn't want to manually sort their inventory and/or an option to use a list (like the inventory is now) for quick access?

**Comment by CrigCrag** [comment]
with this system, would it be possible to change the size/layout of storage containers during the round? (ex. your bag gets damaged and can't store as much)

**Comment by deathride58** [comment]
>If it were to block an unreasonable amount of the screen because of scaled up UI that would really suck
(@Unbelievable-Salmon)

Indeed it would suck! Which is part of why Emo's working on allowing the UI to be moved. For reference, here's how the current static UI in this PR looks with 2x scaling on a 1080p display:
image

This is almost as obtrusive as the modern/default screen layout's chatbox at 2x scaling! This is even worse with larger storage items, but thankfully there's only a handful of jobs that have to worry about those on a regular basis.

When the UI's made movable, this simply won't be an issue for most common storage items, as you'd be able to shunt the storage UI away, like so:
image

Needless to say, this is a *lot* better, even if it means sacrificing a third of the chatbox in the process! Hmeister's idea of allowing the storage UI itself to be rotated would do wonders to accommodate for storage items with weirder bounds (ones that are excessively tall, for instance, which would otherwise be tough to find a good place for at 2x scale), but as they simply don't exist in this PR's current state, that feature is more of a nice-to-have affordance rather than a hard UX necessity at the moment.

**Comment by EmoGarbage404** [comment]
> @EmoGarbage404 Think I might have discovered a bug in this PR, items in tool belts are not displayed properly on the sprites: image
> 
> This is how it looks like in master: image

this is because toolbelts don't have the container configured yet. like i said, it's still heavily WIP and most containers and pretty much nonfunctional.

**Comment by DiaAWAY** [comment]
> > @EmoGarbage404 Think I might have discovered a bug in this PR, items in tool belts are not displayed properly on the sprites: image
> > This is how it looks like in master: image
> 
> this is because toolbelts don't have the container configured yet. like i said, it's still heavily WIP and most containers and pretty much nonfunctional.

Alright, I'll come back in a couple days and take it for a spin then :)

**Comment by EmoGarbage404** [comment]
https://github.com/space-wizards/space-station-14/assets/98561806/e7908e65-d80c-4fa8-9dd6-d8060a4af835

rejoice (defaults to true in options)

**Comment by PJB3005** [comment]
I would give the draggable storage UI an actual "handle" area on one of the sides, so it's obvious that they can click drag.

**Comment by LankLTE** [comment]
If I’m understanding, the ability to drag the storage UI is disabled by default?

**Comment by EmoGarbage404** [comment]
> I would give the draggable storage UI an actual "handle" area on one of the sides, so it's obvious that they can click drag. 

Wdym? The draggable area is the sidebar along the right side with the X button.

**Comment by EmoGarbage404** [comment]
> > I would give the draggable storage UI an actual "handle" area on one of the sides, so it's obvious that they can click drag. 
> 
> Wdym? The draggable area is the sidebar along the left side with the X button.

Also yeah its bound to the hotbar by default

**Comment by LankLTE** [comment]
> Also yeah its bound to the hotbar by default

Cool, so now that I know I’m not completely wrong here, gotta ask… why? Having it be movable is almost objectively better- absolute worst case someone likes it above the hot bar and just… doesn’t move it.

**Comment by SonicHDC** [comment]
Why only 1 inventory shows? Multiple storages are planed?

**Comment by EmoGarbage404** [comment]
READY FOR REVIEW

**Comment by Kit0vras** [comment]
What happens if, during mapping, I put an item in a [filled] backpack in a random place? Is this provided for? Will the item not disappear/an error will not appear? Or do you need to look at how things are laid out in the backpack when filling it and put it where they definitely won't appear?

**Comment by Hmeister-fake** [comment]
I really hope this gets test merged soon, hopefully wont have as much of an uproar of hate as the other inventory reworks. Can't wait to play some inventory tetris as a doctor with a bunch of chem bottles, i love organising.

**Comment by EmoGarbage404** [comment]
> What happens if, during mapping, I put an item in a [filled] backpack in a random place? Is this provided for? Will the item not disappear/an error will not appear? Or do you need to look at how things are laid out in the backpack when filling it and put it where they definitely won't appear?

it'll just auto fill around it. i would be careful though, as filling container with storageFills in mapping mode is an excellent way to cause loading errors.

**Comment by 0x6273** [comment]
Rotation is bugged. The rectangle rotates but not the sprite inside it.

image

image

**Comment by EmoGarbage404** [comment]
> Rotation is bugged. The rectangle rotates but not the sprite inside it.
> 
> image
> 
> image

that was intentional. i didn't want to reduce the recognition of sprites by making them rotate, especially since their names aren't displayed in the ui. obviously it might be beneficial for some specific sprites, but i felt that generally it's better if the item sprite was agnostic to the rotation of the grid. you can see the same design in the original proposal mockups.

**Comment by EmoGarbage404** [comment]
I mean, yes, but it the goal is for a "one size fits most" approach. Yeah, they're gonna overlap slightly, but they're still recognizable. Its not really practical to be adjusting shapes and sprites for 500+ items to make sure they rotate and look perfect. Someone could always add in functionality to do that, but i felt like it was best to go for a generic approach that works with very little extra effort (all of the items were converted with no extra yaml)

**Comment by EmoGarbage404** [comment]
it was discussed at some point being able to have custom sprites for storage and so i might implement that in a later pr if people feel strongly about implementing custom sprites for some of the more commonly used items.

File: Content.Shared/Storage/EntitySystems/SharedStorageSystem.cs [hunk 1]
```diff
@@ -363,17 +421,48 @@ private void OnBoundUIOpen(EntityUid uid, StorageComponent storageComp, BoundUIO
...
+    {
+        // ReSharper disable once ConditionIsAlwaysTrueOrFalseAccordingToNullableAPIContract
+        if (entity.Comp.Container == null)
+            return;
+
+        if (args.Container.ID != StorageComponent.ContainerId)
+            return;
+
+        if (_net.IsServer)
+        {
```
Comment (metalgearsloth) [inline Content.Shared/Storage/EntitySystems/SharedStorageSystem.cs:434]: Why is this wrapped in a server check?

Is this supposed to be an ApplyingState check or something?

**Review by metalgearsloth — DISMISSED** [review]

**Pull request merged**